File: ts/ingest/obsidian.ts
/**
 * Obsidian vault connector — the curation layer. Everything is
 * quality_tier=curated; [[wikilinks]] feed the link graph; vault PRs are the
 * write-back target that makes the corpus compound.
 */

import { readFileSync, readdirSync, statSync } from "node:fs";
import { join, relative, sep } from "node:path";
import { CanonicalDoc } from "../contracts/models.js";
import { extractLinks } from "./common.js";

export function noteId(vaultRoot: string, path: string): string {
  const rel = relative(vaultRoot, path).replace(/\.md$/, "");
  return `obsidian:note:${rel.split(sep).join("/")}`;
}

export function normalizeNote(vaultRoot: string, path: string, tenant = "default"): CanonicalDoc {
  const body = readFileSync(path, "utf-8");
  const docId = noteId(vaultRoot, path);
  const relParts = relative(vaultRoot, path).split(sep);
  let title = relParts[relParts.length - 1]!.replace(/\.md$/, "");
  for (const line of body.split("\n")) {
    if (line.startsWith("# ")) {
      title = line.slice(2).trim();
      break;
    }
  }
  return CanonicalDoc.parse({
    id: docId,
    tenant,
    source: "obsidian",
    title,
    updatedAt: new Date(statSync(path).mtimeMs).toISOString(),
    hierarchy: relParts.slice(0, -1),
    aclGroups: [],
    qualityTier: "curated",
    body,
    links: extractLinks(body, docId),
  });
}

export function walkVault(vaultRoot: string, tenant = "default"): CanonicalDoc[] {
  const files: string[] = [];
  const walk = (dir: string) => {
    for (const entry of readdirSync(dir, { withFileTypes: true })) {
      if (entry.name === ".obsidian" || entry.name === ".trash") continue;
      const full = join(dir, entry.name);
      if (entry.isDirectory()) walk(full);
      else if (entry.name.endsWith(".md")) files.push(full);
    }
  };
  walk(vaultRoot);
  files.sort();
  return files.map((p) => normalizeNote(vaultRoot, p, tenant));
}

File: src/eil/ingest/obsidian.py
"""Obsidian vault connector — the curation layer.

A vault is git-synced markdown; everything in it is quality_tier=curated (a
human wrote or approved it), the top ranking prior. [[wikilinks]] feed the
same link graph as ticket keys. This is also the write-back target: workflows
propose notes as PRs, humans merge, the corpus compounds.
"""

from __future__ import annotations

from datetime import UTC, datetime
from pathlib import Path

from eil.ingest.common import extract_links
from eil.models import CanonicalDoc


def note_id(vault_root: Path, path: Path) -> str:
    rel = path.relative_to(vault_root).with_suffix("")
    return f"obsidian:note:{rel.as_posix()}"


def normalize_note(vault_root: Path, path: Path, tenant: str = "default") -> CanonicalDoc:
    body = path.read_text(encoding="utf-8")
    doc_id = note_id(vault_root, path)
    rel = path.relative_to(vault_root)
    title = path.stem
    for line in body.split("\n"):
        if line.startswith("# "):
            title = line[2:].strip()
            break
    return CanonicalDoc(
        id=doc_id,
        tenant=tenant,
        source="obsidian",
        title=title,
        author=None,
        updated_at=datetime.fromtimestamp(path.stat().st_mtime, tz=UTC),
        hierarchy=list(rel.parts[:-1]),
        acl_groups=[],
        quality_tier="curated",
        body=body,
        links=extract_links(body, doc_id),
    )


def walk_vault(vault_root: Path, tenant: str = "default") -> list[CanonicalDoc]:
    return [
        normalize_note(vault_root, p, tenant)
        for p in sorted(vault_root.rglob("*.md"))
        if ".obsidian" not in p.parts and ".trash" not in p.parts
    ]
